IA 340 Lab 2


In [4]:
import warnings
warnings.simplefilter('ignore')

import pandas as pd

df = pd.read_csv('s3://harrigan-ia340/Arrests.csv') # load the data from s3
df.head()#show the top 5 rows

CB_NO CASE NUMBER             ARREST DATE   RACE       CHARGE 1 STATUTE  \
0  18922700         NaN  06/26/2014 10:51:00 PM  BLACK     725 ILCS 5.0/110-3   
1  18922086         NaN  06/26/2014 12:07:00 AM  BLACK   625 ILCS 5.0/6-303-A   
2  18908465    HX290868  06/05/2014 11:58:00 AM  BLACK   720 ILCS 570.0/402-C   
3  18914257    HX302763  06/14/2014 04:50:00 AM  BLACK  720 ILCS 5.0/12-3-A-1   
4  18911035         NaN  06/09/2014 08:45:00 AM  BLACK                8-4-030   

                                CHARGE 1 DESCRIPTION CHARGE 1 TYPE  \
0                                ISSUANCE OF WARRANT           NaN   
1                       DRIVING ON SUSPENDED LICENSE             M   
2  PCS - POSSESSION - POSS AMT CON SUB EXCEPT (A)(D)             F   
3                        BATTERY - CAUSE BODILY HARM             M   
4                 DRINKING ALCOHOL ON THE PUBLIC WAY           NaN   

  CHARGE 1 CLASS      CHARGE 2 STATUTE                CHARGE 2 DESCRIPTION  \
0              Z                   NaN                                 NaN   
1              A           9-76-160(A)           DISPLAY ST REG-FRONT/REAR   
2              4  720 ILCS 600.0/3.5-A  PCS - DRUG PARAPHERNALIA - POSSESS   
3              A     520 ILCS 5.0/1.22             RESIST/OBSTRUCT OFFICER   
4              L                   NaN                                 NaN   

   ... CHARGE 3 TYPE CHARGE 3 CLASS CHARGE 4 STATUTE CHARGE 4 DESCRIPTION  \
0  ...           NaN            NaN              NaN                  NaN   
1  ...           NaN            NaN              NaN                  NaN   
2  ...           NaN            NaN              NaN                  NaN   
3  ...             M              A              NaN                  NaN   
4  ...           NaN            NaN              NaN                  NaN   

  CHARGE 4 TYPE CHARGE 4 CLASS  \
0           NaN            NaN   
1           NaN            NaN   
2           NaN            NaN   
3           NaN            NaN   
4           NaN            NaN   

                                     CHARGES STATUTE  \
0                         725 ILCS 5.0/110-3 |  |  |   
1            625 ILCS 5.0/6-303-A | 9-76-160(A) |  |   
2   720 ILCS 570.0/402-C | 720 ILCS 600.0/3.5-A |  |   
3  720 ILCS 5.0/12-3-A-1 | 520 ILCS 5.0/1.22 | 72...   
4                                    8-4-030 |  |  |   

                                 CHARGES DESCRIPTION CHARGES TYPE  \
0                        ISSUANCE OF WARRANT |  |  |      |  |  |   
1  DRIVING ON SUSPENDED LICENSE | DISPLAY ST REG-...    M |  |  |   
2  PCS - POSSESSION - POSS AMT CON SUB EXCEPT (A)...   F | M |  |   
3  BATTERY - CAUSE BODILY HARM | RESIST/OBSTRUCT ...  M | M | M |   
4         DRINKING ALCOHOL ON THE PUBLIC WAY |  |  |      |  |  |   

  CHARGES CLASS  
0     Z |  |  |  
1    A | L |  |  
2    4 | A |  |  
3   A | A | A |  
4     L |  |  |  

[5 rows x 24 columns]

In [6]:
df.shape

(699030, 24)

there are 24 columns, and only 1 column is a numerical column
#there are a total of 699030 arrests in the data, and 391,873 of those arrests have only one charge

In [25]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 699030 entries, 0 to 699029
Data columns (total 24 columns):
 #   Column                Non-Null Count   Dtype 
---  ------                --------------   ----- 
 0   CB_NO                 699030 non-null  int64 
 1   CASE NUMBER           440608 non-null  object
 2   ARREST DATE           699030 non-null  object
 3   RACE                  699030 non-null  object
 4   CHARGE 1 STATUTE      699030 non-null  object
 5   CHARGE 1 DESCRIPTION  699030 non-null  object
 6   CHARGE 1 TYPE         549580 non-null  object
 7   CHARGE 1 CLASS        694339 non-null  object
 8   CHARGE 2 STATUTE      307157 non-null  object
 9   CHARGE 2 DESCRIPTION  307157 non-null  object
 10  CHARGE 2 TYPE         175418 non-null  object
 11  CHARGE 2 CLASS        304965 non-null  object
 12  CHARGE 3 STATUTE      170594 non-null  object
 13  CHARGE 3 DESCRIPTION  170594 non-null  object
 14  CHARGE 3 TYPE         59933 non-null   object
 15  CHARGE 3 CLASS   

In [30]:
df=df.rename(columns={'CHARGE 1 DESCRIPTION': 'CHARGE 1 DESC'})
df.head()

CB_NO CASE NUMBER             ARREST DATE   RACE       CHARGE 1 STATUTE  \
0  18922700         NaN  06/26/2014 10:51:00 PM  BLACK     725 ILCS 5.0/110-3   
1  18922086         NaN  06/26/2014 12:07:00 AM  BLACK   625 ILCS 5.0/6-303-A   
2  18908465    HX290868  06/05/2014 11:58:00 AM  BLACK   720 ILCS 570.0/402-C   
3  18914257    HX302763  06/14/2014 04:50:00 AM  BLACK  720 ILCS 5.0/12-3-A-1   
4  18911035         NaN  06/09/2014 08:45:00 AM  BLACK                8-4-030   

                                       CHARGE 1 DESC CHARGE 1 TYPE  \
0                                ISSUANCE OF WARRANT           NaN   
1                       DRIVING ON SUSPENDED LICENSE             M   
2  PCS - POSSESSION - POSS AMT CON SUB EXCEPT (A)(D)             F   
3                        BATTERY - CAUSE BODILY HARM             M   
4                 DRINKING ALCOHOL ON THE PUBLIC WAY           NaN   

  CHARGE 1 CLASS      CHARGE 2 STATUTE                CHARGE 2 DESCRIPTION  \
0              Z                   NaN                                 NaN   
1              A           9-76-160(A)           DISPLAY ST REG-FRONT/REAR   
2              4  720 ILCS 600.0/3.5-A  PCS - DRUG PARAPHERNALIA - POSSESS   
3              A     520 ILCS 5.0/1.22             RESIST/OBSTRUCT OFFICER   
4              L                   NaN                                 NaN   

   ... CHARGE 3 TYPE CHARGE 3 CLASS CHARGE 4 STATUTE CHARGE 4 DESCRIPTION  \
0  ...           NaN            NaN              NaN                  NaN   
1  ...           NaN            NaN              NaN                  NaN   
2  ...           NaN            NaN              NaN                  NaN   
3  ...             M              A              NaN                  NaN   
4  ...           NaN            NaN              NaN                  NaN   

  CHARGE 4 TYPE CHARGE 4 CLASS  \
0           NaN            NaN   
1           NaN            NaN   
2           NaN            NaN   
3           NaN            NaN   
4           NaN            NaN   

                                     CHARGES STATUTE  \
0                         725 ILCS 5.0/110-3 |  |  |   
1            625 ILCS 5.0/6-303-A | 9-76-160(A) |  |   
2   720 ILCS 570.0/402-C | 720 ILCS 600.0/3.5-A |  |   
3  720 ILCS 5.0/12-3-A-1 | 520 ILCS 5.0/1.22 | 72...   
4                                    8-4-030 |  |  |   

                                 CHARGES DESCRIPTION CHARGES TYPE  \
0                        ISSUANCE OF WARRANT |  |  |      |  |  |   
1  DRIVING ON SUSPENDED LICENSE | DISPLAY ST REG-...    M |  |  |   
2  PCS - POSSESSION - POSS AMT CON SUB EXCEPT (A)...   F | M |  |   
3  BATTERY - CAUSE BODILY HARM | RESIST/OBSTRUCT ...  M | M | M |   
4         DRINKING ALCOHOL ON THE PUBLIC WAY |  |  |      |  |  |   

  CHARGES CLASS  
0     Z |  |  |  
1    A | L |  |  
2    4 | A |  |  
3   A | A | A |  
4     L |  |  |  

[5 rows x 24 columns]

I renamed the column Charge 1 Description to Charge 1 Desc

In [32]:
df.isnull().sum() 

CB_NO                        0
CASE NUMBER             258422
ARREST DATE                  0
RACE                         0
CHARGE 1 STATUTE             0
CHARGE 1 DESC                0
CHARGE 1 TYPE           149450
CHARGE 1 CLASS            4691
CHARGE 2 STATUTE        391873
CHARGE 2 DESCRIPTION    391873
CHARGE 2 TYPE           523612
CHARGE 2 CLASS          394065
CHARGE 3 STATUTE        528436
CHARGE 3 DESCRIPTION    528436
CHARGE 3 TYPE           639097
CHARGE 3 CLASS          529072
CHARGE 4 STATUTE        604307
CHARGE 4 DESCRIPTION    604307
CHARGE 4 TYPE           674467
CHARGE 4 CLASS          604508
CHARGES STATUTE              0
CHARGES DESCRIPTION          0
CHARGES TYPE                 0
CHARGES CLASS                0
dtype: int64

There are a lot of missing variables in the charge 2, 3, and 4 columns because not every arrest has more than 1 charge associated with that arrest. I think leaving the dataset to contain the NaN would be important because it shows that there is no data for the other charges.

In [33]:
df[['ARREST DATE']]

ARREST DATE
0       06/26/2014 10:51:00 PM
1       06/26/2014 12:07:00 AM
2       06/05/2014 11:58:00 AM
3       06/14/2014 04:50:00 AM
4       06/09/2014 08:45:00 AM
...                        ...
699025  08/21/2025 06:40:00 PM
699026  08/21/2025 08:02:00 PM
699027  08/21/2025 08:20:00 PM
699028  08/21/2025 03:09:00 PM
699029  08/21/2025 07:02:00 PM

[699030 rows x 1 columns]

In [37]:
df.sort_values(by='CHARGE 1 DESC')

CB_NO CASE NUMBER             ARREST DATE            RACE  \
670529  30409020    JH502819  11/10/2024 02:21:00 PM  WHITE HISPANIC   
640325  18873417    HX221440  04/12/2014 09:20:00 PM  WHITE HISPANIC   
71669   19060017    HY147478  02/11/2015 12:20:00 AM           WHITE   
340201  19509437    JA350972  07/16/2017 11:07:00 PM           WHITE   
512564  19338513    HZ338152  07/06/2016 10:10:00 AM  WHITE HISPANIC   
...          ...         ...                     ...             ...   
541482  19802623    JC241836  04/28/2019 04:50:00 PM  WHITE HISPANIC   
460926  30252607    JG298960  06/13/2023 07:10:00 AM  WHITE HISPANIC   
491697  30334272    JH189288  03/15/2024 08:10:00 PM  WHITE HISPANIC   
662395  30402020    JH473659  10/18/2024 07:57:00 PM  WHITE HISPANIC   
544606  19820331    JC292026  06/04/2019 05:05:00 PM           BLACK   

                 CHARGE 1 STATUTE                           CHARGE 1 DESC  \
670529  720 ILCS 570.0/401-C-7-II          10<15 OBJECTS/PARTS LSD/ANALOG   
640325  720 ILCS 570.0/401-C-7-II          10<15 OBJECTS/PARTS LSD/ANALOG   
71669   720 ILCS 570.0/401-C-7-II          10<15 OBJECTS/PARTS LSD/ANALOG   
340201   720 ILCS 570.0/401-A-7-A                 15<100 GRAMS LSD/ANALOG   
512564   720 ILCS 570.0/401-A-7-A                 15<100 GRAMS LSD/ANALOG   
...                           ...                                     ...   
541482                9-24-040(A)  YIELD TO VEHICLE ENTERING INTERSECTION   
460926                9-24-040(A)  YIELD TO VEHICLE ENTERING INTERSECTION   
491697                9-52-040(C)                   YIELDING RIGHT OF WAY   
662395                9-52-040(C)                   YIELDING RIGHT OF WAY   
544606                9-52-040(C)                   YIELDING RIGHT OF WAY   

       CHARGE 1 TYPE CHARGE 1 CLASS          CHARGE 2 STATUTE  \
670529             F              1      720 ILCS 570.0/401-E   
640325             F              1                       NaN   
71669              F              1        720 ILCS 550.0/4-D   
340201             F              X    720 ILCS 570.0/401-C-2   
512564             F              X  720 ILCS 570.0/402-A-2-A   
...              ...            ...                       ...   
541482           NaN              L      625 ILCS 5.0/3-707-A   
460926           NaN              L        625 ILCS 5.0/6-101   
491697           NaN              L        625 ILCS 5.0/6-101   
662395           NaN              L      625 ILCS 5.0/3-707-A   
544606           NaN              L                       NaN   

                       CHARGE 2 DESCRIPTION  ... CHARGE 3 TYPE CHARGE 3 CLASS  \
670529  PCS-MFG/DEL- NON-NARCOTIC-SCHD I&II  ...             F              4   
640325                                  NaN  ...           NaN            NaN   
71669        CANNABIS - POSSESS 30-500 GRMS  ...           NaN              Z   
340201    PCS - MFG/DEL1<15 GR COCAINE/ANLG  ...             F              3   
512564   PCS - POSSESS 15<100 GRAMS COCAINE  ...             F              2   
...                                     ...  ...           ...            ...   
541482        OPERATE UNINSURED MTR VEHICLE  ...           NaN            NaN   
460926                           UNLICENSED  ...           NaN              P   
491697                           UNLICENSED  ...           NaN              P   
662395        OPERATE UNINSURED MTR VEHICLE  ...           NaN              P   
544606                                  NaN  ...           NaN            NaN   

             CHARGE 4 STATUTE  \
670529   720 ILCS 570.0/402-C   
640325                    NaN   
71669      625 ILCS 5.0/3-707   
340201  720 ILCS 5.0/26-1-A-1   
512564     720 ILCS 550.0/4-E   
...                       ...   
541482                    NaN   
460926                    NaN   
491697                    NaN   
662395                    NaN   
544606                    NaN   

                                     CHARGE 4 DESCRIPTION CHARGE 4 TYP

In [38]:
df[5:10]

CB_NO CASE NUMBER             ARREST DATE            RACE  \
5  18921756    HX318527  06/25/2014 04:35:00 PM           BLACK   
6  18911533    HX297219  06/10/2014 12:15:00 AM  WHITE HISPANIC   
7  18919006    HX312887  06/21/2014 12:30:00 PM           BLACK   
8  18926798         NaN  07/03/2014 12:45:00 AM           BLACK   
9  18914147    HX302600  06/14/2014 12:05:00 AM           BLACK   

          CHARGE 1 STATUTE                             CHARGE 1 DESC  \
5    720 ILCS 5.0/19-4-A-1                CRIM TRESPASS TO RESIDENCE   
6  720 ILCS 5.0/12-3.2-A-2       DOMESTIC BATTERY - PHYSICAL CONTACT   
7                 10-8-515              SOLICITING UNLAWFUL BUSINESS   
8       725 ILCS 5.0/110-3                       ISSUANCE OF WARRANT   
9      720 ILCS 5.0/31-1-A  RESISTING/OBSTRUCT/PC OFF/CORR EMP/FRFTR   

  CHARGE 1 TYPE CHARGE 1 CLASS       CHARGE 2 STATUTE  \
5             M              A                    NaN   
6             M              A  720 ILCS 5.0/12-3.5-A   
7           NaN              L                    NaN   
8           NaN              Z                    NaN   
9             M              A             8-4-010(E)   

                       CHARGE 2 DESCRIPTION  ... CHARGE 3 TYPE CHARGE 3 CLASS  \
5                                       NaN  ...           NaN            NaN   
6             INTERF REPT DOMESTIC VIOLENCE  ...           NaN            NaN   
7                                       NaN  ...           NaN            NaN   
8                                       NaN  ...           NaN            NaN   
9  DISORDERLY CONDUCT - FAIL TO OBEY POLICE  ...           NaN            NaN   

  CHARGE 4 STATUTE CHARGE 4 DESCRIPTION CHARGE 4 TYPE CHARGE 4 CLASS  \
5              NaN                  NaN           NaN            NaN   
6              NaN                  NaN           NaN            NaN   
7              NaN                  NaN           NaN            NaN   
8              NaN                  NaN           NaN            NaN   
9              NaN                  NaN           NaN            NaN   

                                     CHARGES STATUTE  \
5                      720 ILCS 5.0/19-4-A-1 |  |  |   
6  720 ILCS 5.0/12-3.2-A-2 | 720 ILCS 5.0/12-3.5-...   
7                                   10-8-515 |  |  |   
8                         725 ILCS 5.0/110-3 |  |  |   
9              720 ILCS 5.0/31-1-A | 8-4-010(E) |  |   

                                 CHARGES DESCRIPTION CHARGES TYPE  \
5                 CRIM TRESPASS TO RESIDENCE |  |  |    M |  |  |   
6  DOMESTIC BATTERY - PHYSICAL CONTACT | INTERF R...   M | M |  |   
7               SOLICITING UNLAWFUL BUSINESS |  |  |      |  |  |   
8                        ISSUANCE OF WARRANT |  |  |      |  |  |   
9  RESISTING/OBSTRUCT/PC OFF/CORR EMP/FRFTR | DIS...    M |  |  |   

  CHARGES CLASS  
5     A |  |  |  
6    A | A |  |  
7     L |  |  |  
8     Z |  |  |  
9    A | L |  |  

[5 rows x 24 columns]

In [17]:
df[df['RACE'] == 'BLACK']

CB_NO CASE NUMBER             ARREST DATE   RACE  \
0       18922700         NaN  06/26/2014 10:51:00 PM  BLACK   
1       18922086         NaN  06/26/2014 12:07:00 AM  BLACK   
2       18908465    HX290868  06/05/2014 11:58:00 AM  BLACK   
3       18914257    HX302763  06/14/2014 04:50:00 AM  BLACK   
4       18911035         NaN  06/09/2014 08:45:00 AM  BLACK   
...          ...         ...                     ...    ...   
699023  30500202    JJ383899  08/21/2025 03:10:00 PM  BLACK   
699024  30500190    JJ383843  08/21/2025 02:30:00 PM  BLACK   
699025  30500251    JJ115877  08/21/2025 06:40:00 PM  BLACK   
699028  30500233    JJ383899  08/21/2025 03:09:00 PM  BLACK   
699029  30500273         NaN  08/21/2025 07:02:00 PM  BLACK   

               CHARGE 1 STATUTE  \
0            725 ILCS 5.0/110-3   
1          625 ILCS 5.0/6-303-A   
2          720 ILCS 570.0/402-C   
3         720 ILCS 5.0/12-3-A-1   
4                       8-4-030   
...                         ...   
699023    720 ILCS 5.0/24-1.7-A   
699024  720 ILCS 5.0/12-3.2-A-2   
699025      720 ILCS 5.0/12-1-A   
699028   720 ILCS 570.0/401-C-1   
699029                 10-8-526   

                                     CHARGE 1 DESCRIPTION CHARGE 1 TYPE  \
0                                     ISSUANCE OF WARRANT           NaN   
1                            DRIVING ON SUSPENDED LICENSE             M   
2       PCS - POSSESSION - POSS AMT CON SUB EXCEPT (A)(D)             F   
3                             BATTERY - CAUSE BODILY HARM             M   
4                      DRINKING ALCOHOL ON THE PUBLIC WAY           NaN   
...                                                   ...           ...   
699023                      POSS FIREARM REPEAT FEL OFFEN             F   
699024                     DOMESTIC BTRY/PHYSICAL CONTACT             M   
699025                                   ASSAULT - SIMPLE             M   
699028                PCS - MFG/DEL 1<15 GR HEROIN/ANALOG             F   
699029        CTA - TRESPASS ON/W/O PAYING (98-126-1.14C)           NaN   

       CHARGE 1 CLASS        CHARGE 2 STATUTE  \
0                   Z                     NaN   
1                   A             9-76-160(A)   
2                   4    720 ILCS 600.0/3.5-A   
3                   A       520 ILCS 5.0/1.22   
4                   L                     NaN   
...               ...                     ...   
699023              X  720 ILCS 570.0/401-C-1   
699024              A                     NaN   
699025              C                     NaN   
699028              1                     NaN   
699029              L                     NaN   

                       CHARGE 2 DESCRIPTION  ... CHARGE 3 TYPE CHARGE 3 CLASS  \
0                                       NaN  ...           NaN            NaN   
1                 DISPLAY ST REG-FRONT/REAR  ...           NaN            NaN   
2        PCS - DRUG PARAPHERNALIA - POSSESS  ...           NaN            NaN   
3                   RESIST/OBSTRUCT OFFICER  ...             M              A   
4                                       NaN  ...           NaN            NaN   
...                                     ...  ...           ...            ...   
699023  PCS - MFG/DEL 1<15 GR HEROIN/ANALOG  ...             M              A   
699024                                  NaN  ...           NaN            NaN   
699025                                  NaN  ...           NaN            NaN   
699028                                  NaN  ...           NaN            NaN   
699029                                  NaN  ...           NaN            NaN   

           CHARGE 4 STATUTE                      CHARGE 4 DESCRIPTION  \
0                       NaN                                       NaN   
1                       NaN                                       NaN   
2                       NaN                                       NaN   
3                       NaN                                       NaN   
4    

In [ ]:
I sorted the charge description into alphabetical order
I filtered the race of the person being arrested to black, and 490